src/core/tree_types.ml

```diff
@@ -61,12 +61,17 @@ type tree_cond =
 type 'rhs tree =
   | Fail
   (** Empty decision tree, used when there are no rewriting rules. *)
-  | Leaf of (int * (int * int array)) list * 'rhs
-  (** The value [Leaf(m, rhs)] stores the RHS [rhs] of the rewriting rule that
+  | Leaf of (int * (int * int array)) list * 'rhs * int
+  (** Triple [Leaf(m, rhs, x)] stores the RHS [rhs] of the rewriting rule that
```

Comment:
Do you really need x in this type? In eval.ml, can't you use act.xvars_nb instead?

Reply:
But `act` does not have such an attribute, it's a really the rhs as a binder, so we have to ship this `xvars_nb` in the `Leaf`

Reply:
Ah yes it's a better idea to modify the instantiated `'rhs` type to carry this `act.xvars_nb`.